Quotient/e2ee/cryptoutils.cpp

```diff
@@ -0,0 +1,315 @@
+        ctx.get(),
+        encryptedSpan.subspan(static_cast<size_t>(encryptedLength)).data(),
+        &tailLength));
+    Q_ASSUME(tailLength == 0); // Specific to AES CTR
+
+    return encrypted.copyToByteArray(encryptedLength + tailLength);
+}
+
+SslExpected<HkdfKeys> Quotient::hkdfSha256(const QByteArray& key,
+                                           const QByteArray& salt,
+                                           const QByteArray& info)
+{
+    CLAMP_SIZE(saltSize, salt);
+    CLAMP_SIZE(keySize, key);
+    CLAMP_SIZE(infoSize, info);
+
+    HkdfKeys result;
+    const ContextHolder context(EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, nullptr),
+                                &EVP_PKEY_CTX_free);
+
+    CALL_OPENSSL(EVP_PKEY_derive_init(context.get()));
+    CALL_OPENSSL(EVP_PKEY_CTX_set_hkdf_md(context.get(), EVP_sha256()));
+    CALL_OPENSSL(EVP_PKEY_CTX_set1_hkdf_salt(context.get(),
+                                             asCBytes(salt).data(), saltSize));
+    CALL_OPENSSL(EVP_PKEY_CTX_set1_hkdf_key(context.get(), asCBytes(key).data(),
+                                            keySize));
+    CALL_OPENSSL(EVP_PKEY_CTX_add1_hkdf_info(context.get(),
+                                             asCBytes(info).data(), infoSize));
+    size_t outputLength = result.size();
+    CALL_OPENSSL(EVP_PKEY_derive(context.get(), result.data(), &outputLength));
+    if (outputLength != result.size()) {
+        qCCritical(E2EE) << "hkdfSha256: the shared secret is" << outputLength
+                         << "bytes instead of" << result.size();
+        Q_ASSERT(false);
+        return WrongDerivedKeyLength;
+    }
+
+    return result;
+}
+
+SslExpected<QByteArray> Quotient::hmacSha256(byte_view_t<HmacKeySize> hmacKey,
+                                             const QByteArray& data)
+{
+    unsigned int len = SHA256_DIGEST_LENGTH;
+    auto output = zeroedByteArray(SHA256_DIGEST_LENGTH);
+    if (HMAC(EVP_sha256(), hmacKey.data(), hmacKey.size(), asCBytes(data).data(),
+             unsignedSize(data), asWritableCBytes(output).data(), &len)
+        == nullptr) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    return output;
+}
+
+SslExpected<QByteArray> Quotient::aesCtr256Decrypt(const QByteArray& ciphertext,
+                                                   byte_view_t<Aes256KeySize> key,
+                                                   byte_view_t<AesBlockSize> iv)
+{
+    const ContextHolder context(EVP_CIPHER_CTX_new(), &EVP_CIPHER_CTX_free);
+    if (!context) {
+        qCCritical(E2EE)
+            << "aesCtr256Decrypt() failed to create cipher context:"
+            << ERR_error_string(ERR_get_error(), nullptr);
+        Q_ASSERT(context);
+        return ERR_get_error();
+    }
+
+    CLAMP_SIZE(ciphertextSize, ciphertext);
+
+    FixedBuffer<> decrypted(static_cast<size_t>(ciphertextSize),
+                            FixedBufferBase::FillWithZeros);
```

Comment:
Not to myself who added this code in #716: this is not right. `FixedBuffer`s use secure memory, which is currently in a rather short supply (64 KiB) and only meant to store secrets. This code in particular is used to decrypt attachments, and a file of any considerable size now literally leads to an assertion failure/crash.